_posts/2019-12-18-package-lock.md

```diff
@@ -0,0 +1,101 @@
+---
+layout: post
+title: To lock or not to lock
+tags:
+- node.js
+status: publish
+category: node.js
+type: post
+published: true
+---
+
+![Package Lock Files](/images/2019/lock.png)
+
+A few years ago, [npm](https://npmjs.org) introduced the notion of a [`package-lock.json`](https://docs.npmjs.com/files/package-lock.json). The purpose of the file is to provide a manifest that calls out the exact version of every package in your tree, the last time `npm install` was run.  After running `npm install`, you're going to see a message like this:
+
+<pre><code>npm notice created a lockfile as package-lock.json. You should commit this file.
+</code></pre>
+
+While this advice is well intentioned...  it's not always true :) Let's talk about when you want to check this into source control, and when not to.
+
+## How package-lock.json works
+
+The mechanics of `package-lock.json` are simple enough:
+- When you run `npm install` the first time, npm automatically creates the file for you.
+- It has the exact version of every top level and transitive dependency in your `node_modules` folder.
+- When you update the version of a package in your `package.json` and run `npm install`, the `package-lock.json` file will get updated automatically.
+
+If you peek inside one of these files, you're going to get a bunch of non-human-readable gobbledygook like this:
+
+<pre><code class="language-js">
+{
+  "name": "@google-cloud/bigtable",
+  "version": "2.3.0",
+  "lockfileVersion": 1,
+  "requires": true,
+  "dependencies": {
+    "@babel/code-frame": {
+      "version": "7.5.5",
+      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.5.5.tgz",
+      "integrity": "sha512-27d4lZoomVyo51VegxI20xZPuSHusqbQag/ztrBC7wegWoQ1nLREPVSKSW8byhTlzTKyNE4ifaTA6lCp7JjpFw==",
+      "dev": true,
+      "requires": {
+        "@babel/highlight": "^7.0.0"
+      }
+    },
+  ...
+</code></pre>
+
+The next time you run `npm install`, you're (likely) going to get the exact copy of dependencies outlined in the `package-lock.json` file, instead of going through the module resolution dance from the contents of `package.json`.
+
+## Why lock files sound great
+In a world where lock files do not exist - running `npm install` is largely [non-idempotent](https://en.wikipedia.org/wiki/Idempotence).  This means that I can run `npm install` twice, and the results in the `node_modules` folder can be wildly different.  Stop me if you've seen this happen before:
```

Comment:
re: non-idempotent

rsc@ uses the term "reproducible builds" (or in your case, non-reproducible builds). Perhaps consider using the same? I think it's a more approachable definition for people who aren't familiar with idempotency and don't want to go read a wiki page.

re: example

Another way of phrasing this that might make this a lot scarier for people:

- Run locally, get crypto@1.2.3 (is perfectly safe)
- Deploy to prod, get crypto@1.2.4 (has a major security flaw - oh no!)

----------------

Both of the above are just discussion points - feel free to resolve if you're happy with it as-is.

Reply:
This is a really good call.  I just like having an excuse to use the word idempotent.  It's like 90% of the reason I work with APIs.